### ISSUE TITLE
Choosing a development model

### ISSUE THREAD

**Issue by MartijnBraam** [postmarketOS/pmbootstrap#382]
After the discussion in the chat. We currently use a development model where we push everything to master and have that semi-stable and then use feature/bugfix branches for everything.

I propose we use gitflow which we already do for 80% and make a `develop` branch and make that the default branch. The documentation for gitflow is http://nvie.com/posts/a-successful-git-branching-model/

The changes from the current process:

- We merge every PR to `develop` instead of `master`
- After every release cycle there is a single merge commit to `master` from `develop`
- Versions should be tagged on the `master` branch

The major part of development; the PR's would still be the same. We should also decide on what the release cycle for pmbootstrap would be to `master`.

One way would be to make a fully rolling release with a `unstable` branch (develop) and a `stable` branch (master) and longer supported versions which are the tags. The linux kernel is basically doing a similar release cycle with the weekly RC releases.

The basic rules for gitflow:

- Master should be tested and stable
- Develop might break but is generally stable and hopefully breakage would be detected before releasing to master
- The feature branches should be functionally independent and should only contain a single feature
- Seperate commits inside the feature branch might not build but develop should always build (which is currently tested by travis)
- If master breaks then a `hotfix/[name]` branch is made from master instead of develop and this is merged again to both `master` and `develop`
- Probably more things I forgot

**Comment by opendata26** [comment]
+ 1 but i dont think the feature freeze ise that appropriate to pmos

**Comment by ollieparanoid** [comment]
I see both advantages and disadvantages in this model for `pmbootstrap`, and I'm very interested what everyone else thinks!

Advantages:
* git bisect probably becomes a lot simpler
* more stable experience for first time users

Disadvantages:
* a little more effort for new users (which is against the *make it as easy as possible principle*)
* not sure if that make sense for the `aports` folder (see splitting it off in #383)
* I am unsure how the test cycle before a release should look like. No one can test all devices, and we're a loose collective of free-time hackers, which will probably not have time to test all devices just before the release window. Testing specific PRs for specific devices works better as I see it.
* I think all developers, which want to contribute switch to `develop`, and then they do not get any of the stability benefits, right?
* much more overhead to actually make the releases, we could put that time in actually fixing bugs/implementing features


*Okay, judging from the list I just wrote down, I'm more against it, than for it. But if everyone else thinks it's a good idea, then we should do as the community decides!*

**Comment by MartijnBraam** [comment]
yeah the master branch is more for end users (and would probably the branch being published/packaged to pypi), for new developers there isn't really any overhead since gitflow and githubflow are the same in that part, just the PR target branch changes from master to develop, most developers would be on the `develop` branch yeah.

To explain the testing more: For merges from feature branches to develop the unit tests for that feature should be run and pass. This means that all seperate units in the `develop` branch should be stable and working. For the release from `develop` to `master` you would run the bigger integration tests.

You mention the overhead for testting all devices but this would only apply to the `pmb` code in the future if the aports stuff is split from this repository, and all the code in `pmb` should have coverage from the test suite.

**Comment by ollieparanoid** [comment]
Okay, that sounds very reasonable. Having such automated integration tests would be wonderful. You have convinced me, I think this is a good idea now :+1:

@opendata26: I don't think, that the `gitflow` works without freezing features until the next release. That's how you get the stability after all. But if we make a release every 1 or 2 weeks, then people would get the shiny new stuff still relatively fast!

**Comment by drebrez** [comment]
I'm also a fan of the `successful-git-branching-model `, we also use it at work and works pretty good.
Another advantage I can see, it's also the scalability in the branch permissions, very few will have the permission to write to master, and some more on the develop branch.

**Comment by ollieparanoid** [comment]
One thing I forgot is, that when we keep `pmbootstrap` and `aports` in one repository, *and* use `gitflow`, we will need to put some extra thought in the upcoming binary repo process. So we might want to do the `aports` split before applying this development model.
